subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/DefaultBasePluginExtension.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.plugins.internal;
+
+import org.gradle.api.Project;
+import org.gradle.api.file.DirectoryProperty;
+import org.gradle.api.plugins.BasePluginExtension;
+
+public class DefaultBasePluginExtension implements BasePluginExtension {
+    private final DirectoryProperty buildDirectory;
+    private final DirectoryProperty distsDirectory;
+    private final DirectoryProperty libsDirectory;
+
+    private String distsDirName;
+    private String libsDirName;
+    private String archivesBaseName;
+
+    public DefaultBasePluginExtension(Project project) {
+        this.buildDirectory = project.getLayout().getBuildDirectory();
+        this.archivesBaseName = project.getName();
+
+        this.distsDirName = "distributions";
+        this.distsDirectory = project.getObjects().directoryProperty();
+        distsDirectory.convention(buildDirectory.dir(distsDirName));
+
+        this.libsDirName = "libs";
+        this.libsDirectory = project.getObjects().directoryProperty();
+        libsDirectory.convention(buildDirectory.dir(libsDirName));
```

Comment:
This should be moved up into the plugin instead of being set in the constructor here.